client.go

```diff
@@ -0,0 +1,377 @@
+	var response payload.QueryResponse
+
+	request, err := c.makeJSONRequest("POST", "/query", query)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := c.doJSONRequest(request, &response); err != nil {
+		return nil, err
+	}
+
+	return &response, nil
+}
+
+// Get returns the body of the specified blob.
+// If `readRange` is non-nil, Get will attempt returns that section of the blob.
+func (c *Client) Get(blobID string, readRange *Range) (io.ReadCloser, error) {
+	if readRange != nil {
+		return &rangeReader{BlobID: blobID, Client: c, RangeStart: readRange.Start, RangeEnd: readRange.End}, nil
+	}
+
+	req, err := c.makeJSONRequest("GET", fmt.Sprintf("/blob/%s", blobID), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	redirectLocation, err := c.doWithRedirect(req)
+	if err != nil {
+		return nil, err
+	}
+
+	req.URL = redirectLocation
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Body, nil
+}
+
+func (c *Client) GetMetadata(blobID string) (payload.BlobMeta, error) {
+	req, err := c.makeJSONRequest("GET", fmt.Sprintf("/blob/%s/metadata", blobID), nil)
+	if err != nil {
+		return payload.BlobMeta{}, err
+	}
+
+	var response payload.GetMetadataResponse
+	if err := c.doJSONRequest(req, &response); err != nil {
+		return payload.BlobMeta{}, err
+	}
+
+	if response.Metadata == nil {
+		return payload.BlobMeta{}, fmt.Errorf("get meta: blob '%s' not found", blobID)
+	}
+
+	return *response.Metadata, nil
+}
+
+// Delete deletes a blob from the cluster.
+func (c *Client) Delete(blobID string) error {
+	req, err := c.makeJSONRequest("DELETE", fmt.Sprintf("/blob/%s", blobID), nil)
+	if err != nil {
+		return err
+	}
+
+	redirectLocation, err := c.doWithRedirect(req)
+	if err != nil {
+		return err
+	}
+
+	req.URL = redirectLocation
+
+	var response payload.MessageResponse
+	if err := c.doJSONRequest(req, &response); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Push creates a blob with the provided body and metadata to the cluster.
+// If the body is nil, the blob is created empty.
+func (c *Client) Push(body io.ReadCloser, meta payload.BlobMeta) (string, error) {
+	return c.pushInternal("/blob", body, meta)
+}
+
+// UpdateBlob updates the entirety of a blob's contents and metadata at once.
+func (c *Client) UpdateBlob(blobID string, body io.ReadCloser, meta payload.BlobMeta) error {
+	_, err := c.pushInternal(fmt.Sprintf("/blob/%s", blobID), body, meta)
+	return err
+}
+
+// UpdateMeta updates exclusively the blob metadata.
+func (c *Client) UpdateMeta(blobID string, meta payload.BlobMeta) error {
+	var response payload.MessageResponse
+	req, err := c.makeJSONRequest("PUT", fmt.Sprintf("/blob/%s/metadata", blobID), &meta)
+	if err != nil {
+		return err
+	}
+
+	redirectLocation, err := c.doWithRedirect(req)
+	if err != nil {
+		return err
+	}
+
+	// Go doesn't like keeping headers between requests, we'll rebuild from scratch.
```

Comment:
I'd have to play with it some more, but there seems to be _some_ weirdness around which parts of the request are kept around when re-doing it and which are not. I remember adding this to fix a hard to debug issue, but I'd have to double-check to see if it still is relevant

Reply:
Update: this is because _both_ server components read the request body, and go doesn't know how to keep the body around for the second request. Will update the comment